### PR TITLE
Generate variations from new attributes screen

### LINKED ISSUE #8111: Generate all variations from new attribute screen
### Description
After creating attributes for a new variable product created in the mobile app, we should allow users to generate all variations for those attributes.

This task is about adding the possibility to request generating all variations in this screen:

Discussed internally: p1672842556925899-slack-C02KUCFCSFP

### PR DESCRIPTION
**Pull request by atorresveiga**
Closes: #8111

### Why
After creating attributes for a new variable product, we wanted to add the generate all variation option along with the generate random variation that was already present on the screen.

### Description
This PR adds the necessary changes to support generate-all-variation and generate-single-variation from the success attributes created screen.

### Testing instructions
Follow these steps to reach the Attributes created screen
1. Add a new variable product or open a variable product without attributes or variations
2. Tap on Add variations
3. Tap on an attribute
4. Tap on several attribute options
5. Tap on next on the top of the screen once you are happy with the options added
6. Tap on next again to reach the Attributes created screen

TC1
1. On the Attributes created screen, tap on generate variation
2. Check that a Bottom sheet opens with the options 1) Add new variation, 2) Generate all variations
3. Tap on Generate all variations
4. Check that the generate all variation dialog is shown
5. Tap on Ok
6. Check that if the operation succeeded, the new variations are added to the product

TC2
1. On the Attributes created screen, tap on generate variation
2. Check that a Bottom sheet opens with the options 1) Add new variation, 2) Generate all variations
3. Tap on Generate new variation
4. Check that if the operation succeeded, a new variation is added to the product

TC3
1. Repeat TC1 and TC2 but remove the internet connection this time before generating the variations
2. Check that an error message is shown

TC4
1. Try to generate more than 100 variations
2. Check that an error message is shown

### Images/gif

https://user-images.githubusercontent.com/18119390/211961500-67155fbe-5195-441a-9ad7-06574739baab.mp4

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt [hunk 1]
```diff
@@ -415,23 +424,86 @@ class ProductDetailViewModel @Inject constructor(
...
+            } ?: triggerEvent(ShowGenerateVariationsError.NetworkError)
+        }
+    }
+
+    fun onAddAllVariationsClicked() {
+        val remoteProductId = viewState.productDraft?.remoteId ?: return
+        tracker.track(AnalyticsEvent.PRODUCT_VARIATION_GENERATION_REQUESTED)
+        launch {
+            attributeListViewState = attributeListViewState.copy(isFetchingVariations = true)
+            variationRepository.getAllVariations(remoteProductId)
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/ProductDetailViewModel.kt]: ⚙️ WDYT about moving "fetching all variations" part into `GenerateVariationCandidates` or into a different use case (e.g. `CreateVariationCandidates`?) that will use `GenerateVariationCandidates` ?

This way we don't have to repeat fetching all variations, and there are fewer chances we'll forget about it in the future. Also, it'll be easier to communicate with VM about failed variations fetching progress (p1673532821172819-slack-C02KUCFCSFP)

**Comment by ParaskP7** [comment]
FYI @atorresveiga @wzieba that I moved the milestone of this PR to `12.2` in order to progress with the `12.1` code freeze today. Do let me know if you still want this PR to be landed in `12.1`. 👍

**Comment by wzieba** [comment]
It's okay to move it to a later release @ParaskP7 . I'll remove the milestone for now as we focus on other priorities.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/domain/GenerateVariationCandidates.kt [hunk 2]
```diff
@@ -20,12 +20,12 @@ class GenerateVariationCandidates @Inject constructor(
...
     }
 
-    operator fun invoke(product: Product): List<VariationCandidate> {
+    suspend operator fun invoke(product: Product): List<VariationCandidate> {
         if (product.type != ProductType.VARIABLE.value) {
             return emptyList()
         }
 
-        val existingVariations = variationRepository.getProductVariationList(product.remoteId)
+        val existingVariations = variationRepository.getAllVariations(product.remoteId)
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/domain/GenerateVariationCandidates.kt:28]: I've changed the `getProductVariationList` -> `getAllVariations`. Now we will use the API as the source of truth and don't need to remember to fetch variations before using them.
We already discussed about how trustworthy `getAllVariations` is, but I think the refactoring of this function is out of the scope of this PR

**Comment by atorresveiga** [comment]
Thanks for your review, @wzieba! I agreed with the changes proposed for moving the fetching variations function into the use case so we can forget about that when using `GenerateVariationCandidates`. I think we are ready for a second review now. 😁

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/domain/GenerateVariationCandidates.kt (hunk 2 above)
Comment (wzieba) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/domain/GenerateVariationCandidates.kt:28]: Good points here 👍 . I've just added an issue to revisit `getAllVariations`: https://github.com/woocommerce/woocommerce-android/issues/8357

**Review by wzieba — APPROVED** [review]
Looks good to me, thanks for adding many unit tests!

**Review by wzieba — APPROVED** [review]

**Pull request merged**